### PR TITLE
imx: add 6.1 as testing kernel

### PR DESCRIPTION
**Pull request by Gateworks**
This adds support for linux 6.1 testing kernel for the imx target.

This has been tested on:
- Gateworks imx6qdl gateworks_ventana (cortexa9)

Hoping for:
- reviews from @pepe2k, @ynezz, @pepe2k, @Leo-PL
- testing from @Leo-PL for pico-pi-imx7

### CODE REVIEW

File: target/linux/imx/config-6.1 [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+CONFIG_XZ_DEC_ARM=y
+CONFIG_XZ_DEC_ARMTHUMB=y
+CONFIG_XZ_DEC_BCJ=y
+CONFIG_ZBOOT_ROM_BSS=0
+CONFIG_ZBOOT_ROM_TEXT=0
+CONFIG_ZLIB_DEFLATE=y
+CONFIG_ZLIB_INFLATE=y
+CONFIG_ZSTD_COMMON=y
+CONFIG_ZSTD_COMPRESS=y
+CONFIG_ZSTD_DECOMPRESS=y
```
Comment (namiltd) [inline target/linux/imx/config-6.1:476]:
```
# CONFIG_DRM_FSL_LDB is not set

**Review by namiltd — COMMENTED** [review]
missing: # CONFIG_DRM_FSL_LDB is not set

**Comment by Gateworks** [comment]
> missing: # CONFIG_DRM_FSL_LDB is not set

Thanks - not sure how that got missed in my testing.

I'm new to github pull requests. I've read that I have to push a new patch on top that fixes issues in the pull request (which sounds strange as I would rather rebase and have a clean set of patches). I see github has updated the pull request automatically but I'm not clear how to kick off the review cycle again.

Any help explaining the github pull workflow when you need to make changes is greatly apprecaited!

**Comment by namiltd** [comment]
You can rollback your local repository, make corrections and then force push.

**Comment by namiltd** [comment]
There are still bugs in the tests: https://github.com/openwrt/openwrt/actions/runs/7563432589

**Comment by Leo-PL** [comment]
@Gateworks `CONFIG_DRM_DW_HDMI_GP_AUDIO` is missing from `config-default`. To debug that, run
```
make target/linux/clean
make target/linux/compile -j1 V=s
```
and then note the symbols `syncconfig` was asking to be filled out manually, then add them. This should fix it. You may take a look at my branch for pico-pi-imx7 camera to look for the symbols, but I think it was intertwined with stuff needed for V4L2 too.

Also CC @pepe2k and @ynezz

**Comment by Leo-PL** [comment]
Just checked on the CI:
`# CONFIG_VIDEO_IMX_MIPI_CSIS is not set` is needed too in `target/linux/generic/config-6.1`

**Comment by Gateworks** [comment]
yes, I saw that it failed on 3 targets all for different kernel
configs needed to be undefined:
VIDEO_IMX_MIPI_CSIS layerscape, armv7, 6.1
VIDEO_ROCKCHIP_ISP1 rockchip, armv8
VIDEO_SUN6I_CSI sunxi, cortexa53

I suppose I need to figure out which patch causes those and put the
undef in the right patch. The first one for example is caused by
adding ***@***.*** to
KernelPackage/video-core which could potentially affect any 6.1
targets but that can't be the cause of the others as they are not 6.1
builds.

I assume it's normal to run into this from time to time and CI check
the targets for you instead of me checking every single target? Or is
there a simple (yet very time/build intensive way) for me to test each
target/subtarget with
CONFIG_ALL_KMODS=y, CONFIG_LINUX_6_1=y and CONFIG_TESTING_KERNEL=y?

What is the proper way to determine where in the config-6.1 things
that need to be disabled need to go?

**Comment by xback** [comment]
I'll try to give this series a spin (Monday) on a gw5400 with gstreamer and the hw ip (coda etc)

**Comment by Leo-PL** [comment]
Okay, tested on iMX7D Pico Pi, on both old and new kernel. You can add my Tested-by: tag.
I noticed one (probably unrelated) issue, that uboot-imx expects `swig` package from the host, but there is no check for that at build start. This only happens on `cortexa9`.

**Review by Leo-PL — APPROVED** [review]

**Comment by Leo-PL** [comment]
As for review - +1, although I probably shouldn't be reviewing changes made partially by myself.

**Comment by Gateworks** [comment]
> I'll try to give this series a spin (Monday) on a gw5400 with gstreamer and the hw ip (coda etc)

Any feedback? I'm not sure how to get 6 reviewers for PR's - doesn't seem like there are a lot of reviewers out there

**Comment by Gateworks** [comment]
> There are still bugs in the tests: https://github.com/openwrt/openwrt/actions/runs/7563432589

All the tests run now with the latest update - do you have time for a review? Any pointers on getting some additional reviewers?

**Comment by Ansuel** [comment]
@Gateworks what happen to the first commit? looks like 2 commit got squashed? Please check

**Comment by Gateworks** [comment]
> @Gateworks what happen to the first commit? looks like 2 commit got squashed? Please check

I don't believe the first 2 got squashed: https://github.com/openwrt/openwrt/pull/14419/commits

I did add additional patches to the first commit from @Leo-PL (vs dropping them in my original PR commit)

**Comment by Ansuel** [comment]
Can we restore the original commit from leo and add an additional commit from you refreshing the patch?

**Comment by Gateworks** [comment]
> Can we restore the original commit from leo and add an additional commit from you refreshing the patch?

the original commit was mine 'imx: copy patches 5.15 to 6.1' and refreshed only arch/arm/boot/dts/imx6dl-wandboard.dts and dropped the others as they didn't cleanly refresh (if I recall). After my original post I added patches from Leo's github I was pointed to that refreshed the dropped patches so I merged and did a force push. I don't have the original anymore.

**Comment by Leo-PL** [comment]
> > Can we restore the original commit from leo and add an additional commit from you refreshing the patch?
> 
> the original commit was mine 'imx: copy patches 5.15 to 6.1' and refreshed only arch/arm/boot/dts/imx6dl-wandboard.dts and dropped the others as they didn't cleanly refresh (if I recall). After my original post I added patches from Leo's github I was pointed to that refreshed the dropped patches so I merged and did a force push. I don't have the original anymore.

You may still have that in `git reflog`. And the original is still in https://github.com/Leo-PL/openwrt/commits/pico-pi-imx7-camera-6.1/ if that helps

**Comment by Gateworks** [comment]
> > > Can we restore the original commit from leo and add an additional commit from you refreshing the patch?
> > 
> > 
> > the original commit was mine 'imx: copy patches 5.15 to 6.1' and refreshed only arch/arm/boot/dts/imx6dl-wandboard.dts and dropped the others as they didn't cleanly refresh (if I recall). After my original post I added patches from Leo's github I was pointed to that refreshed the dropped patches so I merged and did a force push. I don't have the original anymore.
> 
> You may still have that in `git reflog`. And the original is still in https://github.com/Leo-PL/openwrt/commits/pico-pi-imx7-camera-6.1/ if that helps

@Ansuel did you want me to push again with my original commit (https://github.com/Leo-PL/openwrt/commit/d1be506d555ce079d3f52066d7de578cd133a32e) and Leo's follow-on that refreshes the remaining (previously dropped) patches (https://github.com/Leo-PL/openwrt/commit/a481d48ccb262b99ac0505248ff0b22fde99257b)?

Seems to me to be better to have squashed those two like I did but I think I see the issue... my commit message should at least be squashed as it currently shows:

    imx: copy patches 5.15 to 6.1
    
    copy 5.15 patches, refreshing as needed
    
    Signed-off-by: Tim Harvey <tharvey@gateworks.com>
    
    imx: 6.1: refresh remaining patches
    
    Only the Apalis Ixora actually required manual touchups.
    
    Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

**Comment by Leo-PL** [comment]
I'm fine with either, splitting them would probably make for a cleaner Git history. It shouldn't cause build failures, too if someone wants to run `git bisect` later.

**Comment by Ansuel** [comment]
Yes please split them. I don't need the very original commit just the thing split for a better git history...

**Comment by Gateworks** [comment]
> Yes please split them. I don't need the very original commit just the thing split for a better git history...

sounds like the right thing to do. I just updated.

**Comment by Gateworks** [comment]
same patchset (just split) now fails with missing symbols again? why is this so fragile?

**Comment by Leo-PL** [comment]
Isn't the PR actually automatically rebased in CI before running tests? It would seem so, because there were some changes to Raspberry Pi support.

**Comment by Gateworks** [comment]
> Isn't the PR actually automatically rebased in CI before running tests? It would seem so, because there were some changes to Raspberry Pi support.

ya, I guess so. Looks like bcm2712 subtarget was added to bcm27xx now requiring VIDEO_RASPBERRYPI_PISP_BE to be undefined when CONFIG_MEDIA_PLATFORM_DRIVERS=y

I assume its just a matter of that series getting accepted before this one. I'll update again

**Comment by Leo-PL** [comment]
+1.

**Comment by xback** [comment]
Fully works as seen on TV :+1: 

Also all other hw ip blocks seem to work correctly:

```
[    9.297395] imx-ipuv3 2400000.ipu: IPUv3H probed
[    9.303318] imx-ipuv3 2800000.ipu: IPUv3H probed
...
[    9.345825] dwhdmi-imx 120000.hdmi: Detected HDMI TX controller v1.30a with HDCP (DWC HDMI 3D TX PHY)
...
[   10.230489] coda 2040000.vpu: Firmware code revision: 46076
[   10.236099] coda 2040000.vpu: Initialized CODA960.
[   10.240919] coda 2040000.vpu: Firmware version: 3.1.1
[   10.246213] coda 2040000.vpu: coda-jpeg-encoder registered as video8
[   10.252919] coda 2040000.vpu: coda-jpeg-decoder registered as video9
[   10.259474] coda 2040000.vpu: coda-video-encoder registered as video10
[   10.266226] coda 2040000.vpu: coda-video-decoder registered as video11
```

[Full bootlog](https://pastebin.com/raw/CfJyynfG)

Tested also using gstreamer, both encoding & decoding on the VPU

@Ansuel 
Feel free to merge if you also consent

**Comment by Leo-PL** [comment]
Yay! I can now rebase my camera support for iMX7 Pico Pi on top of that.

**Pull request merged**